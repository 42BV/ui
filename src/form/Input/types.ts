export type Position = 'left' | 'right';
export type Mask = (string | RegExp)[];
