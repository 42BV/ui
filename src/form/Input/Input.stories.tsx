import React from 'react';
import { action } from '@storybook/addon-actions';

import { FieldInput, Input, JarbInput } from './Input';
import {
  FieldFormElementDependencies,
  FinalForm,
  JarbFormElementDependencies
} from '../../story-utils';
import { Alert, InputGroupText } from 'reactstrap';
import { Card } from '../../core/Card/Card';
import { AddonIcon } from '../AddonIcon/AddonIcon';
import { Tooltip } from '../../core/Tooltip/Tooltip';
import { Icon } from '../../core/Icon';
import { Button } from '../../core/Button/Button';

function isSuperman(value: string) {
  return value === 'superman' ? undefined : 'Not "superman"';
}

export default {
  title: 'Form/Input',

  decorators: [
    (Story) => (
      <>
        <Alert color="warning" className="mb-4">
          <p>
            To be able to use Input, you have to add react-text-mask to your
            dependencies:
          </p>
          <code>npm install --save react-text-mask</code>
        </Alert>
        <Story />
      </>
    )
  ],

  parameters: {
    component: Input
  }
};

const BasicStory = () => {
  return (
    <Card className="m2">
      <Input
        id="firstName"
        label="First name"
        placeholder="Please enter your first name"
        onChange={(value) => action(`You entered ${value}`)}
      />
    </Card>
  );
};

export const Basic = {
  render: BasicStory,
  name: 'basic'
};

const WithAddonStory = () => {
  return (
    <Card className="m2">
      <Input
        id="firstName"
        label="First name"
        placeholder="Please enter your first name"
        onChange={(value) => action(`onChange: ${value}`)}
        addon={<AddonIcon icon="face" />}
      />

      <Input
        id="lastName"
        label="Last name"
        placeholder="Please enter your last name"
        onChange={(value) => action(`onChange: ${value}`)}
        addon={<AddonIcon icon="face" />}
        color="success"
      />

      <Input
        id="magicNumber"
        label="Magic number"
        placeholder="Please enter the magic number"
        onChange={(value) => action(`onChange: ${value}`)}
        addon={<InputGroupText>Nr</InputGroupText>}
      />

      <Input
        id="magicNumber"
        label="Magic number"
        placeholder="Please enter the magic number"
        onChange={(value) => action(`onChange: ${value}`)}
        addon={<InputGroupText>Try 42</InputGroupText>}
        addonPosition="right"
      />

      <Input
        id="addonButton"
        label="Addon as a button"
        placeholder="Please click on my addon button"
        onChange={(value) => action(`onChange: ${value}`)}
        addon={<Button onClick={() => alert('Yippy')}>Click me</Button>}
      />
    </Card>
  );
};

export const WithAddon = {
  render: WithAddonStory,
  name: 'with addon'
};

const WithoutPlaceholderStory = () => {
  return (
    <Card className="m2">
      <Input
        id="firstName"
        label="First name"
        onChange={(value) => action(`You entered ${value}`)}
      />
    </Card>
  );
};

export const WithoutPlaceholder = {
  render: WithoutPlaceholderStory,
  name: 'without placeholder'
};

const MaskStory = () => {
  return (
    <Card className="m2">
      <Input
        id="zipcode"
        label="Zipcode"
        placeholder="Please enter your zipcode"
        onChange={(value) => action(`You entered ${value}`)}
        mask={[/[1-9]/, /[1-9]/, /[1-9]/, /[1-9]/, ' ', /[A-z]/, /[A-z]/]}
      />
      <p>
        Look
        <a href="https://github.com/text-mask/text-mask/blob/master/componentDocumentation.md#mask">
          here
        </a>
        for more example on how to use mask
      </p>
    </Card>
  );
};

export const Mask = {
  render: MaskStory,
  name: 'mask'
};

const InvisibleLabelStory = () => {
  return (
    <Card className="m2">
      <Input
        label="First name"
        hiddenLabel={true}
        placeholder="Please enter your first name"
        onChange={(value) => action(`You entered ${value}`)}
      />
    </Card>
  );
};

export const InvisibleLabel = {
  render: InvisibleLabelStory,
  name: 'invisible label'
};

const WithCustomLabelStory = () => {
  return (
    <Card className="m2">
      <Input
        id="firstName"
        label={
          <div className="d-flex justify-content-between">
            <span>First name</span>
            <Tooltip
              className="ms-1"
              content="Your first name is on your birth certificate"
            >
              <Icon icon="info" />
            </Tooltip>
          </div>
        }
        placeholder="Please enter your first name"
        onChange={(value) => action(`You entered ${value}`)}
      />
    </Card>
  );
};

export const WithCustomLabel = {
  render: WithCustomLabelStory,
  name: 'with custom label'
};

const FieldStory = () => {
  return (
    <>
      <FieldFormElementDependencies />
      <FinalForm>
        <FieldInput
          name="firstName"
          jarb={{ validator: 'Hero.name', label: 'First name' }}
          validators={[isSuperman]}
          id="firstName"
          label="First name"
          placeholder="Please enter your first name"
        />
      </FinalForm>
    </>
  );
};

export const Field = {
  render: FieldStory,
  name: 'field'
};

const JarbStory = () => {
  return (
    <>
      <JarbFormElementDependencies />
      <FinalForm>
        <JarbInput
          name="firstName"
          jarb={{ validator: 'Hero.name', label: 'First name' }}
          validators={[isSuperman]}
          id="firstName"
          placeholder="Please enter your first name"
        />
      </FinalForm>
    </>
  );
};

export const Jarb = {
  render: JarbStory,
  name: 'jarb'
};
