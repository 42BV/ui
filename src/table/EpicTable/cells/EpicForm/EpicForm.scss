.epic-table-form {
  &,
  &-cell {
    white-space: nowrap;
    overflow: hidden;
    display: flex;
    align-items: center;
    flex-grow: 1;
  }

  form {
    display: flex;
  }

  &--odd {
    .epic-table-form-cell {
      // Must be transparent otherwise the shadow of the left / right
      // columns are negated.
      background-color: rgba($gray-400, 0.3);
    }
  }

  .date-time-input .rdt {
    position: static;

    .rdtPicker {
      top: unset;
      left: unset;
    }
  }
}
