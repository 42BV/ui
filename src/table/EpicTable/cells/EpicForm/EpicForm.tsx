import React from 'react';
import classNames from 'classnames';
import { AnyObject, Form, FormProps } from 'react-final-form';
import { omit, pick } from 'lodash';
import { FormApi } from 'final-form';
import { AutoSave } from '../../../../form/AutoSave/AutoSave';

interface Props<FormValues> extends FormProps<FormValues> {
  /**
   * The id of the form element.
   */
  id: string;

  /**
   * The content of the cell.
   */
  children: React.ReactNode;

  /**
   * The width of the cell.
   */
  width: number;

  /**
   * The height of the cell.
   */
  height: number;

  /**
   * Optionally whether or not the form should be submitted on blur automatically.
   */
  submitOnChange?: boolean;
}

// Props that will be injected by the EpicTable.
interface InjectedProps {
  odd: boolean;
}

/**
 * The EpicForm is used inside of a EpicRow to render content in.
 */
export function EpicForm<FormValues>(props: Props<FormValues>) {
  const { id, children, width, height, submitOnChange, ...rest } = props;
  const { odd } = pick(rest, ['odd']) as InjectedProps;
  const formProps = omit(rest, ['odd']) as FormProps;

  const classes = classNames('epic-table-form border-bottom p-1', {
    'epic-table-form--odd': odd
  });

  return (
    <Form {...formProps}>
      {({
        handleSubmit,
        form
      }: {
        handleSubmit: (
          event?: Partial<
            Pick<React.SyntheticEvent, 'preventDefault' | 'stopPropagation'>
          >
        ) => Promise<AnyObject | undefined> | undefined;
        form: FormApi<FormValues>;
      }) => (
        <form
          id={id}
          onSubmit={handleSubmit}
          onReset={form.reset}
          className={classes}
          style={{
            minWidth: width,
            width,
            height
          }}
        >
          {submitOnChange ? (
            <AutoSave
              onSave={form.submit}
              initialValues={formProps.initialValues}
            />
          ) : null}
          {children}
        </form>
      )}
    </Form>
  );
}
