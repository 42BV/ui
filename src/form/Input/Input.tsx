import React, { ChangeEvent } from 'react';
import RTMaskedInput from 'react-text-mask';
import { FormGroup, Input as RSInput, InputGroup, Label } from 'reactstrap';
import { InputType } from 'reactstrap/lib/Input';
import withJarb from '../withJarb/withJarb';

import Addon, { Props as AddonProps } from './Addon/Addon';

import { Mask } from './types';
import { Color } from '../types';

interface BaseProps {
  /**
   * The placeholder of the form element.
   */
  placeholder?: string | null;

  /**
   * The value that the form element currently has.
   */
  value?: string;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: string) => void;

  /**
   * Optional callback for when the form element is blurred.
   */
  onBlur?: () => void;

  /**
   * Optional callback for when the form element is focused.
   */
  onFocus?: () => void;

  /**
   * Optional type of the input, default to `text`.
   */
  type?: InputType;

  /**
   * Optional mask of the input.
   *
   * @see https://text-mask.github.io/text-mask/
   */
  mask?: Mask;

  /**
   * Optional addon to display to the left or right of the input
   * element.
   */
  addon?: AddonProps;

  /**
   * A ref to the actual input, can be used to focus the element.
   */
  innerRef?: React.Ref<HTMLInputElement>;

  /**
   * Optionally the error message to render.
   */
  error?: React.ReactNode;

  /**
   * Optionally the color of the FormGroup.
   */
  color?: Color;

  /**
   * Whether or not the form element is currently valid.
   */
  valid?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
}

interface WithoutLabel extends BaseProps {
  id?: string;
  label?: never;
}

interface WithLabel extends BaseProps {
  /**
   * The id of the form element.
   */
  id: string;

  /**
   * The label of the form element.
   */
  label: string;
}

export type Props = WithoutLabel | WithLabel;

/**
 * Input is a basic form element which allows the user to enter text.
 *
 * Supports addons and masks.
 */
export default function Input(props: Props) {
  const {
    placeholder,
    value,
    onChange,
    onBlur,
    onFocus,
    innerRef,
    type = 'text',
    error,
    color,
    valid,
    mask,
    addon,
    className = ''
  } = props;

  const inputProps = {
    id: 'id' in props ? props.id : undefined,
    type,
    valid,
    invalid: valid === false ? true : undefined,
    value,
    innerRef,
    placeholder: placeholder ? placeholder : '',
    onFocus,
    onChange: (event: ChangeEvent<HTMLInputElement>) =>
      onChange(event.target.value),
    onBlur
  };

  // Cannot pass a reference when the input requires a mask
  const { innerRef: _, ...withoutRef } = inputProps;

  const input =
    mask === undefined ? (
      <RSInput key="input" {...inputProps} />
    ) : (
      <RTMaskedInput
        key="mask"
        mask={mask}
        {...withoutRef}
        render={reactStrapInput}
      />
    );

  let content = input;

  if (addon !== undefined) {
    const { position } = addon;
    const addonComponent = <Addon key="button" color={color} {...addon} />;

    const inputGroupContent =
      position === 'left' ? [addonComponent, input] : [input, addonComponent];

    content = <InputGroup>{inputGroupContent}</InputGroup>;
  }

  return (
    <FormGroup className={className} color={color}>
      {'label' in props && props.label ? (
        <Label for={props.id}>{props.label}</Label>
      ) : null}
      {content}
      {error}
    </FormGroup>
  );
}

export function reactStrapInput(
  ref: (input: HTMLInputElement) => void,
  props: Record<string, any>
) {
  return <RSInput innerRef={ref} {...props} />;
}

/**
 * Variant of the Input which can be used in a Jarb context.
 */
export const JarbInput = withJarb<string, string, Props>(Input);
