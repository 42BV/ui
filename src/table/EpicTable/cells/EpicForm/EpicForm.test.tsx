import React from 'react';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import '@testing-library/jest-dom';

import { EpicForm } from './EpicForm';
import { FieldInput } from '../../../../form/Input/Input';
import userEvent from '@testing-library/user-event';
import { setConstraints } from '@42.nl/jarb-final-form';

describe('Component: EpicForm', () => {
  function setup({ submitOnChange }: { submitOnChange?: boolean }) {
    setConstraints({
      Test: {
        test: {
          javaType: 'String',
          name: 'test',
          required: true
        }
      }
    });

    const onSubmitSpy = jest.fn();

    const { container } = render(
      <EpicForm
        id="test"
        width={1950}
        height={52}
        initialValues={{}}
        onSubmit={onSubmitSpy}
        submitOnChange={submitOnChange}
      >
        <FieldInput id="testField" name="testField" label="Test" />
        <button type="reset">Reset</button>
      </EpicForm>
    );

    return { container, onSubmitSpy };
  }

  test('ui', async () => {
    expect.assertions(1);
    const { container } = setup({});
    await act(() => {
      fireEvent.focus(screen.getByRole('textbox'));
    });
    expect(container).toMatchSnapshot();
  });

  describe('events', () => {
    it('should call submit on changes when submitOnChange is true', async () => {
      expect.assertions(2);

      const { onSubmitSpy } = setup({ submitOnChange: true });

      fireEvent.focus(screen.getByRole('textbox'));
      fireEvent.change(screen.getByRole('textbox'), {
        target: { value: 'test' }
      });
      fireEvent.blur(screen.getByRole('textbox'));

      await waitFor(() => {
        expect(onSubmitSpy).toHaveBeenCalledTimes(1);
      });
    });

    it('should when resetting the form call reset on the the react-final-form Form', async () => {
      expect.assertions(1);

      setup({});

      await userEvent.type(screen.getByLabelText('Test'), 'test');
      await act(() => {
        fireEvent.click(screen.getByText('Reset'));
      });

      expect(screen.getByLabelText('Test')).toHaveValue('');
    });
  });
});
