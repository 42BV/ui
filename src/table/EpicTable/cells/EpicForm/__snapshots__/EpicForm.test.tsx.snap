// Vitest Snapshot v1

exports[`Component: EpicForm > ui 1`] = `
<div>
  <form
    class="epic-table-form border-bottom p-1"
    id="test"
    style="min-width: 1950px; width: 1950px; height: 52px;"
  >
    <div
      class="mb-3"
      color=""
    >
      <label
        class="form-label"
        for="testField"
      >
        Test
      </label>
      <input
        class="form-control"
        defaultvalue=""
        id="testField"
        type="text"
      />
    </div>
    <button
      type="reset"
    >
      Reset
    </button>
  </form>
</div>
`;
